Add tags to aws ecs task

ezmac opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Adds support for tags to ecs_task.  Partially addresses #51122 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ecs_task

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

ezmac committed:
Add tags to ecs_task

remove older reference to credentials

ezmac committed:
uncomment shell commands to enable/disable account settings

ezmac committed:
Fix documentation and pep8 issues

willthames commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -66,6 +66,11 @@
         required: false
         version_added: 2.8
         choices: ["EC2", "FARGATE"]
+    task_tags:
```
Can this just be tags for consistency with every other AWS module?

ezmac replied:
It should have been.  My syntax highlighter got confused and I did as well.

Thanks for your time reviewing and the suggestions you gave.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -243,14 +253,21 @@ def run_task(self, cluster, task_definition, overrides, count, startedBy, launch
+            if (self.ecs_task_long_format_enabled()):
+                params['tags'] = ansible_dict_to_boto3_tag_list(tags, 'key', 'value')
+            else:
+                self.module.fail_json(msg="Account does not support long format task arns")
```
I get that long format task ARNS might be a prerequisite for tags but might be better say e.g.

> "Cannot set task tags: long format task arns are required to set tags"

willthames commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -339,6 +370,9 @@ def main():
         module.fail_json(msg='botocore needs to be version 1.8.4 or higher to use launch type')
 
+    if module.params['task_tags'] and not service_mgr.ecs_api_handles_tags():
+        module.fail_json(msg='botocore needs to be version 1.12.46 or higher to use tags')
```
Probably nicer to use 

```
msg=missing_required_lib("botocore >= 1.12.46", reason="to use tags")
```

ezmac committed:
fix review items for ecs_task tags

use missing_required_lib for tags
change fail_json message to suggested message
switch from task_tags to tags for consisitency

ezmac committed:
Add import for missing_required_lib function

ezmac commented:
I also moved the check for account support to the same block as the check for boto support so that there's only one place an error is defined.  if you'd rather have it the other way, please let me know.

willthames commented:
Looks good now @ezmac - I'll run this through the test suite in the next couple of days and it should be good to merge.

willthames commented on test/integration/targets/ecs_cluster/playbooks/roles/ecs_cluster/tasks/main.yml:
```diff
@@ -754,10 +759,69 @@
+        state: present
+        name: awscli
+
+    - name: disable taskLongArnFormat
```
This can be expressed far more healthily (i.e. you can see what has gone wrong because you don't have to use no_log) as:

```
      command: aws ecs put-account-setting --name taskLongArnFormat --value disabled
      environment:
        AWS_ACCESS_KEY_ID: "{{ aws_access_key }}"
        AWS_SECRET_ACCESS_KEY: "{{ aws_secret_key }}"
        AWS_SESSION_TOKEN: "{{ security_token }}"
        AWS_DEFAULT_REGION: "{{ aws_region }}"
```

willthames replied:
I'll probably just push up my fix for this shortly, so don't worry too much about it :)

willthames committed:
Tidy put-account-setting tasks and add permission

Using `environment` and `command` rather than `shell` avoids the
need for `no_log` and means that people can fix the problem

willthames commented:
I've added a couple of commits that make tests pass

There are some comprehensive changes to the testing policies due to the limit of 10 policies per group - so need to consolidate policies to get back under 10.

ezmac commented:
Well, I may have messed this PR up trying to fix conflicts.  Sorry about that.  I'll see if I can get it cleaned up again.  Until then, feel free to ignore.

ezmac committed:
update version added for ecs_task tags

ezmac commented:
there I fixed it?  I'll try to check back in a bit now that the 2.8 release has (hopefully) calmed down.  Let me know if there are any changes needed; I'm not using 2.8 in my daily life yet.

willthames commented:
Test suite fails for me again, I think I'll need to totally refactor the ECS test suite which will take a few days, I'll probably split out all the test suite changes into a new PR and then this will sit on top.

The conversion from runme.sh to a role using pip/virtualenv is what's needed, similar to the test suite changes in https://github.com/ansible/ansible/pull/57324 - happy to do that work, it'll just take me a few days

ezmac commented:
No problems here.  It's hard for me to keep up with the pace at which y'all move doing this part time.  Thanks for all the time you've spent cleaning up my work :)

willthames commented:
I've rebased this on #57716 - this one should be held from merge until that one has got through.

willthames reviewed (changes requested):
Hold on merge until #57716 merged

ezmac commented:
@willthames Would you be willing to pick this back up or should I do the work to get it merge worthy again?

willthames commented:
@ezmac I've rebased but haven't yet run a test suite on it - I'll try and do that later on

ezmac committed:
fix tests after removal of ansible_facts from ecs_service_info, add delay when service is still draining

ezmac committed:
Add documentation for sanity tests

jillr reviewed (approved):
@ezmac looks like an unresolved merge conflict got pushed in hacking/aws_config/testing_policies/security-policy.json, I'm happy to merge once that's cleaned up.

ezmac commented:
Sorry about that.  Let me know if the way I cleaned it up is problematic.

jillr commented:
Looks right to me, tests pass locally, thanks @ezmac!
